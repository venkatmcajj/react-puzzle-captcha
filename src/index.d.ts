import React, { ReactNode } from 'react';

export interface IverifyProp {
  /**
   * @description   canvaswidth
   * @default       320
   */
  width?: number;
  /**
   * @description   canvasheight
   * @default       160
   */
  height?: number;
  /**
   * @description   slider side length
   * @default       42
   */
  l?: number;
  /**
   * @description   slider radius
   * @default       9
   */
  r?: number;
  /**
   * @description   Is it visible
   * @default       true
   */
  visible?: boolean;
  /**
   * @description   slider text
   * @default       Swipe right to fill the puzzle
   */
  text?: string | ReactNode;
  /**
   * @description   refresh button icon, It is the url address of the icon
   * @default       -
   */
  refreshIcon?: string;
  /**
   * @description   The url address used to get random pictures
   * @default       https://picsum.photos/${id}/${width}/${height}, specific reference https://picsum.photos/, Just implement a similar interface
   */
  imgUrl?: string;
  /**
   * @description   Verification successful callback
   * @default       ():void => {}
   */
  onSuccess?: VoidFunction;
  /**
   * @description   Validation failure callback
   * @default       ():void => {}
   */
  onFail?: VoidFunction;
  /**
   * @description   Callback on refresh
   * @default       ():void => {}
   */
  onRefresh?: VoidFunction;
}

declare const verify: React.FC<IverifyProp>;

export { verify };
