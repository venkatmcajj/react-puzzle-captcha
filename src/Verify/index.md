### Basic use:

```tsx
import React from 'react';
import { Verify } from 'react-slider-verify';

export default () => {
  return <Verify />;
};
```

### set width and height:

```tsx
import React from 'react';
import { Verify } from 'react-slider-verify';

export default () => {
  return <Verify width={330} height={80} />;
};
```

### Set slider side length and radius:

```tsx
import React from 'react';
import { Verify } from 'react-slider-verify';

export default () => {
  return <Verify width={320} height={160} l={28} r={5} />;
};
```

### Set success, failure, callback when refreshing:

```tsx
import React from 'react';
import { Verify } from 'react-slider-verify';

export default () => {
  return (
    <Verify
      width={320}
      height={160}
      onSuccess={() => alert('success')}
      onFail={() => alert('fail')}
      onRefresh={() => alert('refresh')}
    />
  );
};
```

### User-Defined Validation Logic:

The component exposes `onCustomverify` method, and accepts `verify` object as input parameter, we can control `spliced` and `verified` attributes to control whether the verification is successful, that is, the return value of the function must contain `spliced` and `verified` two Boolean property object

```tsx
import React from 'react';
import { Verify } from 'react-slider-verify';

export default () => {
  const handleCustomverify = (verify) => {
    console.log(verify, Math.abs(left - destX) < 5);
    const { destX, left, spliced, verified } = verify;
    return {
      spliced: Math.abs(left - destX) < 5,
      verified,
    };
  };
  return (
    <Verify
      width={320}
      height={160}
      onCustomverify={handleCustomverify}
      onSuccess={() => alert('success')}
      onFail={() => alert('fail')}
      onRefresh={() => alert('refresh')}
    />
  );
};
```

### Dynamically set show/hide:

```tsx
import React, { useState } from 'react';
import { Verify } from 'react-slider-verify';

export default () => {
  const [visible, setVisible] = useState(false);
  const show = () => {
    setVisible(true);
  };
  const hide = () => {
    setVisible(false);
  };
  const style = {
    display: 'inline-block',
    marginRight: '20px',
    marginBottom: '20px',
    width: '100px',
    padding: '5px 20px',
    color: '#fff',
    textAlign: 'center',
    cursor: 'pointer',
    background: '#1991FA',
  };
  return (
    <>
      <div onClick={show} style={style}>
        Show
      </div>
      <div onClick={hide} style={style}>
        Hide
      </div>
      <Verify
        width={320}
        height={160}
        visible={visible}
        onSuccess={() => alert('success')}
        onFail={() => alert('fail')}
        onRefresh={() => alert('refresh')}
      />
    </>
  );
};
```

<API></API>
