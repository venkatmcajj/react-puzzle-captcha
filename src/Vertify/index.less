.vertifyWrap {
  position: relative;

  .block {
    position: absolute;
    left: 0;
    top: 0;
    cursor: pointer;
    cursor: grab;
  }

  .block:active {
    cursor: grabbing;
  }

  .sliderContainer {
    position: relative;
    text-align: center;
    width: 310px;
    height: 40px;
    line-height: 40px;
    margin-top: 15px;
    background: #f7f9fa;
    color: #45494c;
    border: 1px solid #e4e7eb;
  }

  .sliderContainer_active .slider {
    height: 38px;
    top: -1px;
    border: 1px solid #486cd6;
  }

  .sliderContainer_active .sliderMask {
    height: 38px;
    border-width: 1px;
  }

  .sliderContainer_success .slider {
    height: 38px;
    top: -1px;
    border: 1px solid #0db87f;
    background-color: #0ca14a !important;
  }

  .sliderContainer_success .sliderMask {
    height: 38px;
    border: 1px solid #0db87f;
    background-color: #d2f4ef;
  }

  .sliderContainer_success .sliderIcon {
    background-position: 0 -26px !important;
  }

  .sliderContainer_fail .slider {
    height: 38px;
    top: -1px;
    border: 1px solid #f57a7a;
    background-color: #f57a7a !important;
  }

  .sliderContainer_fail .sliderMask {
    height: 38px;
    border: 1px solid #f57a7a;
    background-color: #fce1e1;
  }

  .sliderContainer_fail .sliderIcon {
    top: 14px;
    background-position: 0 -82px !important;
  }

  .sliderContainer_active .sliderText,
  .sliderContainer_success .sliderText,
  .sliderContainer_fail .sliderText {
    display: none;
  }

  .sliderMask {
    position: absolute;
    left: 0;
    top: 0;
    height: 40px;
    border: 0 solid #486cd6;
    background: #d1e9fe;
  }

  .slider {
    position: absolute;
    top: 0;
    left: 0;
    width: 40px;
    height: 40px;
    background: #fff;
    box-shadow: 0 0 3px rgba(0, 0, 0, 0.3);
    transition: background 0.2s linear;
    cursor: pointer;
    cursor: grab;
  }

  .slider:active {
    cursor: grabbing;
  }

  .slider:hover {
    background: #486cd6;
  }

  .sliderIcon {
    font-size: 18px;
    color: #000;
  }

  .slider:hover .sliderIcon {
    color: #fff;
  }

  .refreshIcon {
    position: absolute;
    right: 5px;
    top: 5px;
    width: 30px;
    height: 30px;
    cursor: pointer;
    background-size: 32px;
  }

  .loadingContainer {
    position: absolute;
    left: 0;
    top: 0;
    width: 310px;
    height: 155px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    font-size: 14px;
    color: #45494c;
    z-index: 2;
    background: #edf0f2;
  }

  .loadingIcon {
    width: 32px;
    height: 32px;
    margin-bottom: 10px;
    background: url(https://raw.githubusercontent.com/venkatmcajj/react-slider-verify/master/docs/icon12.png);
    background-size: 32px;
    animation: loading-icon-rotate 0.8s linear infinite;
  }

  @keyframes loading-icon-rotate {
    from {
      transform: rotate(0);
    }
    to {
      transform: rotate(360deg);
    }
  }
}
