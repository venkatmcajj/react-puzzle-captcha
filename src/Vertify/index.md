### Basic use:

```tsx
import React from 'react';
import { Vertify } from 'react-slider-vertify';

export default () => {
  return <Vertify />;
};
```

### set width and height:

```tsx
import React from 'react';
import { Vertify } from 'react-slider-vertify';

export default () => {
  return <Vertify width={330} height={80} />;
};
```

### Set slider side length and radius:

```tsx
import React from 'react';
import { Vertify } from 'react-slider-vertify';

export default () => {
  return <Vertify width={320} height={160} l={28} r={5} />;
};
```

### Set success, failure, callback when refreshing:

```tsx
import React from 'react';
import { Vertify } from 'react-slider-vertify';

export default () => {
  return (
    <Vertify
      width={320}
      height={160}
      onSuccess={() => alert('success')}
      onFail={() => alert('fail')}
      onRefresh={() => alert('refresh')}
    />
  );
};
```

### User-Defined Validation Logic:

The component exposes `onCustomVertify` method, and accepts `vertify` object as input parameter, we can control `spliced` and `verified` attributes to control whether the verification is successful, that is, the return value of the function must contain `spliced` and `verified` two Boolean property object

```tsx
import React from 'react';
import { Vertify } from 'react-slider-vertify';

export default () => {
  const handleCustomVertify = (vertify) => {
    console.log(vertify, Math.abs(left - destX) < 5);
    const { destX, left, spliced, verified } = vertify;
    return {
      spliced: Math.abs(left - destX) < 5,
      verified,
    };
  };
  return (
    <Vertify
      width={320}
      height={160}
      onCustomVertify={handleCustomVertify}
      onSuccess={() => alert('success')}
      onFail={() => alert('fail')}
      onRefresh={() => alert('refresh')}
    />
  );
};
```

### Dynamically set show/hide:

```tsx
import React, { useState } from 'react';
import { Vertify } from 'react-slider-vertify';

export default () => {
  const [visible, setVisible] = useState(false);
  const show = () => {
    setVisible(true);
  };
  const hide = () => {
    setVisible(false);
  };
  const style = {
    display: 'inline-block',
    marginRight: '20px',
    marginBottom: '20px',
    width: '100px',
    padding: '5px 20px',
    color: '#fff',
    textAlign: 'center',
    cursor: 'pointer',
    background: '#1991FA',
  };
  return (
    <>
      <div onClick={show} style={style}>
        Show
      </div>
      <div onClick={hide} style={style}>
        Hide
      </div>
      <Vertify
        width={320}
        height={160}
        visible={visible}
        onSuccess={() => alert('success')}
        onFail={() => alert('fail')}
        onRefresh={() => alert('refresh')}
      />
    </>
  );
};
```

<API></API>
