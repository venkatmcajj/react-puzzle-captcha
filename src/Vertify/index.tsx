import React, { useRef, useState, useEffect, ReactNode, memo } from 'react';
import { getRandomNumberByRange, sum, square } from './tool';
import './index.less';

interface VertifyType {
  spliced: boolean;
  verified: boolean; // Simply verify the drag trajectory, when it is zero, it means that there is no fluctuation up and down the Y axis, which may be non-human operation
  left: number; // Move position of the slider
  destX: number; // The target position of the slider
}

interface IVertifyProp {
  /**
   * @description   canvaswidth
   * @default       320
   */
  width?: number;
  /**
   * @description   canvasheight
   * @default       160
   */
  height?: number;
  /**
   * @description   slider side length
   * @default       42
   */
  l?: number;
  /**
   * @description   slider radius
   * @default       9
   */
  r?: number;
  /**
   * @description   Is it visible
   * @default       true
   */
  visible?: boolean;
  /**
   * @description   slider text
   * @default       Swipe right to fill the puzzle
   */
  text?: string | ReactNode;
  /**
   * @description   refresh button icon, It is the url address of the icon
   * @default       -
   */
  refreshIcon?: string;
  /**
   * @description   The url address used to get random pictures
   * @default       https://picsum.photos/${id}/${width}/${height}, specific referencehttps://picsum.photos/, Just implement a similar interface
   */
  imgUrl?: string;
  /**
   * @description   Callback when dragging the slider, The parameter is the dragging distance of the current slider
   * @default       (l: number):void => {}
   */
  onDraw?: (l: number) => {};
  /**
   * @description   User's custom validation logic
   * @default       (arg: VertifyType) => VertifyType
   */
  onCustomVertify?: (arg: VertifyType) => VertifyType;
  /**
   * @description   Verification successful callback
   * @default       ():void => {}
   */
  onSuccess?: VoidFunction;
  /**
   * @description   Validation failure callback
   * @default       ():void => {}
   */
  onFail?: VoidFunction;
  /**
   * @description   Callback on refresh
   * @default       ():void => {}
   */
  onRefresh?: VoidFunction;
}

export default memo(
  ({
    width = 320,
    height = 160,
    l = 42,
    r = 9,
    imgUrl,
    text,
    refreshIcon = 'https://raw.githubusercontent.com/venkatmcajj/react-slider-verify/master/docs/icon12.png',
    visible = true,
    onDraw,
    onCustomVertify,
    onSuccess,
    onFail,
    onRefresh,
  }: IVertifyProp) => {
    const [isLoading, setLoading] = useState(false);
    const [sliderLeft, setSliderLeft] = useState(0);
    const [sliderClass, setSliderClass] = useState('sliderContainer');
    const [textTip, setTextTip] = useState(text);
    const canvasRef = useRef<any>(null);
    const blockRef = useRef<any>(null);
    const imgRef = useRef<any>(null);
    const isMouseDownRef = useRef<boolean>(false);
    const trailRef = useRef<number[]>([]);
    const originXRef = useRef<number>(0);
    const originYRef = useRef<number>(0);
    const xRef = useRef<number>(0);
    const yRef = useRef<number>(0);
    const PI = Math.PI;
    const L = l + r * 2 + 3; // The actual side length of the slider

    const drawPath = (
      ctx: any,
      x: number,
      y: number,
      operation: 'fill' | 'clip',
    ) => {
      ctx.beginPath();
      ctx.moveTo(x, y);
      ctx.arc(x + l / 2, y - r + 2, r, 0.72 * PI, 2.26 * PI);
      ctx.lineTo(x + l, y);
      ctx.arc(x + l + r - 2, y + l / 2, r, 1.21 * PI, 2.78 * PI);
      ctx.lineTo(x + l, y + l);
      ctx.lineTo(x, y + l);
      ctx.arc(x + r - 2, y + l / 2, r + 0.4, 2.76 * PI, 1.24 * PI, true);
      ctx.lineTo(x, y);
      ctx.lineWidth = 2;
      ctx.fillStyle = 'rgba(255, 255, 255, 0.7)';
      ctx.strokeStyle = 'rgba(255, 255, 255, 0.7)';
      ctx.stroke();
      ctx.globalCompositeOperation = 'destination-over';
      operation === 'fill' ? ctx.fill() : ctx.clip();
    };

    const getRandomImgSrc = () => {
      return (
        imgUrl ||
        `https://picsum.photos/id/${getRandomNumberByRange(
          0,
          1084,
        )}/${width}/${height}`
      );
    };

    const createImg = (onload: VoidFunction) => {
      const img = new Image();
      img.crossOrigin = 'Anonymous';
      img.onload = onload;
      img.onerror = () => {
        (img as any).setSrc(getRandomImgSrc()); // Reload other images when image loading fails
      };

      (img as any).setSrc = (src: string) => {
        const isIE = window.navigator.userAgent.indexOf('Trident') > -1;
        if (isIE) {
          // IE browser cannot cross domain through img.crossOrigin, use ajax to get image blob and then convert it to dataURL for display
          const xhr = new XMLHttpRequest();
          xhr.onloadend = function (e: any) {
            const file = new FileReader(); // FileReader only supports IE10+
            file.readAsDataURL(e.target.response);
            file.onloadend = function (e) {
              img.src = e?.target?.result as string;
            };
          };
          xhr.open('GET', src);
          xhr.responseType = 'blob';
          xhr.send();
        } else img.src = src;
      };

      (img as any).setSrc(getRandomImgSrc());
      return img;
    };

    const draw = (img: HTMLImageElement) => {
      const canvasCtx = canvasRef.current.getContext('2d');
      const blockCtx = blockRef.current.getContext('2d');
      // Create puzzle shapes at random positions
      xRef.current = getRandomNumberByRange(L + 10, width - (L + 10));
      yRef.current = getRandomNumberByRange(10 + r * 2, height - (L + 10));
      drawPath(canvasCtx, xRef.current, yRef.current, 'fill');
      drawPath(blockCtx, xRef.current, yRef.current, 'clip');

      // draw into picture
      canvasCtx.drawImage(img, 0, 0, width, height);
      blockCtx.drawImage(img, 0, 0, width, height);

      // Extract the slider and bring it to the far left
      const y1 = yRef.current - r * 2 - 1;
      const ImageData = blockCtx.getImageData(xRef.current - 3, y1, L, L);
      blockRef.current.width = L;
      blockCtx.putImageData(ImageData, 0, y1);
    };

    const initImg = () => {
      const img = createImg(() => {
        setLoading(false);
        draw(img);
      });
      imgRef.current = img;
    };

    const reset = () => {
      const canvasCtx = canvasRef.current.getContext('2d');
      const blockCtx = blockRef.current.getContext('2d');
      // reset style
      setSliderLeft(0);
      setSliderClass('sliderContainer');
      blockRef.current.width = width;
      blockRef.current.style.left = 0 + 'px';

      // clear canvas
      canvasCtx.clearRect(0, 0, width, height);
      blockCtx.clearRect(0, 0, width, height);

      // reload image
      setLoading(true);
      imgRef.current.setSrc(getRandomImgSrc());
    };

    const handleRefresh = () => {
      reset();
      typeof onRefresh === 'function' && onRefresh();
    };

    const verify = () => {
      const arr = trailRef.current; // The moving distance of the y-axis when dragging
      const average = arr.reduce(sum) / arr.length;
      const deviations = arr.map((x) => x - average);
      const stddev = Math.sqrt(deviations.map(square).reduce(sum) / arr.length);
      const left = parseInt(blockRef.current.style.left);
      return {
        spliced: Math.abs(left - xRef.current) < 10,
        verified: stddev !== 0, // Simply verify the drag trajectory, when it is zero, it means that there is no fluctuation up and down the Y axis, which may be non-human operation
        left,
        destX: xRef.current,
      };
    };

    const handleDragStart = function (e: any) {
      originXRef.current = e.clientX || e.touches[0].clientX;
      originYRef.current = e.clientY || e.touches[0].clientY;
      isMouseDownRef.current = true;
    };

    const handleDragMove = (e: any) => {
      if (!isMouseDownRef.current) return false;
      e.preventDefault();
      const eventX = e.clientX || e.touches[0].clientX;
      const eventY = e.clientY || e.touches[0].clientY;
      const moveX = eventX - originXRef.current;
      const moveY = eventY - originYRef.current;
      if (moveX < 0 || moveX + 38 >= width) return false;
      setSliderLeft(moveX);
      const blockLeft = ((width - 40 - 20) / (width - 40)) * moveX;
      blockRef.current.style.left = blockLeft + 'px';

      setSliderClass('sliderContainer sliderContainer_active');
      trailRef.current.push(moveY);
      onDraw && onDraw(blockLeft);
    };

    const handleDragEnd = (e: any) => {
      if (!isMouseDownRef.current) return false;
      isMouseDownRef.current = false;
      const eventX = e.clientX || e.changedTouches[0].clientX;
      if (eventX === originXRef.current) return false;
      setSliderClass('sliderContainer');
      const { spliced, verified } = onCustomVertify
        ? onCustomVertify(verify())
        : verify();
      if (spliced) {
        if (verified) {
          setSliderClass('sliderContainer sliderContainer_success');
          typeof onSuccess === 'function' && onSuccess();
        } else {
          setSliderClass('sliderContainer sliderContainer_fail');
          setTextTip('请再试一次');
          reset();
        }
      } else {
        setSliderClass('sliderContainer sliderContainer_fail');
        typeof onFail === 'function' && onFail();
        setTimeout(reset.bind(this), 1000);
      }
    };

    useEffect(() => {
      if (visible) {
        imgRef.current ? reset() : initImg();
      }
    }, [visible]);

    return (
      <div
        className="vertifyWrap"
        style={{
          width: width + 'px',
          margin: '0 auto',
          display: visible ? '' : 'none',
        }}
        onMouseMove={handleDragMove}
        onMouseUp={handleDragEnd}
        onTouchMove={handleDragMove}
        onTouchEnd={handleDragEnd}
      >
        <div className="canvasArea">
          <canvas ref={canvasRef} width={width} height={height}></canvas>
          <canvas
            ref={blockRef}
            className="block"
            width={width}
            height={height}
            onMouseDown={handleDragStart}
            onTouchStart={handleDragStart}
          ></canvas>
        </div>
        <div
          className={sliderClass}
          style={{
            pointerEvents: isLoading ? 'none' : 'auto',
            width: width + 'px',
          }}
        >
          <div className="sliderMask" style={{ width: sliderLeft + 'px' }}>
            <div
              className="slider"
              style={{ left: sliderLeft + 'px' }}
              onMouseDown={handleDragStart}
              onTouchStart={handleDragStart}
            >
              <div className="sliderIcon">&rarr;</div>
            </div>
          </div>
          <div className="sliderText">{textTip}</div>
        </div>
        <div
          className="refreshIcon"
          onClick={handleRefresh}
          style={{ backgroundImage: `url(${refreshIcon})` }}
        ></div>
        <div
          className="loadingContainer"
          style={{
            width: width + 'px',
            height: height + 'px',
            display: isLoading ? '' : 'none',
          }}
        >
          <div className="loadingIcon"></div>
          <span>Loading...</span>
        </div>
      </div>
    );
  },
);
