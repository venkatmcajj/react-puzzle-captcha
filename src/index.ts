export { default as Vertify } from './Vertify';
