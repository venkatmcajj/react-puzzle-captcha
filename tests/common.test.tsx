import React from "react";
import "jest-canvas-mock";
import renderer from "react-test-renderer";
import { verify } from "../src";
import fetchmock from "jest-fetch-mock";
fetchmock.dontMock();
describe("Common render", () => {
  it("should render correctly without crash", () => {
    const tree = renderer.create(<verify />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
