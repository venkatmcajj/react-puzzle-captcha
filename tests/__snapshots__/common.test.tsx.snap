// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common render should render correctly without crash 1`] = `
<div
  style={
    {
      "position": "relative",
    }
  }
>
  <div
    className="stdropdown-container"
  >
    <div
      className="stdropdown-input stsearch-box"
      onClick={[Function]}
    >
      <input
        onChange={[Function]}
        value=""
      />
      <div
        className="stdropdown-tools"
      >
        <div
          className="stdropdown-tool"
        >
          <svg
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
